File: Insert_before_a_given_node.java
public void insertBeforeNode(Node[] head, Node beforeNode, int data) {
    if (beforeNode == null) {
        System.out.println("The node to insert before cannot be null.");
        return;
    }

    Node newNode = new Node(data);

    if (head[0] == beforeNode) {
        newNode.link = head[0];
        head[0] = newNode;
    } else {
        Node current = head[0];
        while (current != null && current.link != beforeNode) {
            current = current.link;
        }

        if (current == null) {
            System.out.println("Node not found. Node not inserted.");
            return;
        }

        newNode.link = beforeNode;
        current.link = newNode;
    }
}

File: Insert_before_a_given_node.c
void insertBeforeNode(NODE* head, NODE beforeNode, int data) {
    if (beforeNode == NULL) {
        printf("The node to insert before cannot be NULL.\n");
        return;
    }

    NODE newNode = createNode(data);

    if (*head == beforeNode) {
        newNode->link = *head;
        *head = newNode;
    } else {
        NODE current = *head;
        while (current != NULL && current->link != beforeNode) {
            current = current->link;
        }

        if (current == NULL) {
            printf("Node not found. Node not inserted.\n");
            return;
        }

        newNode->link = beforeNode;
        current->link = newNode;
    }
}
